rohmu/transfer_pool.py

```diff
@@ -0,0 +1,220 @@
+                raise InvalidTransferError("Trying to access transfer instance already returned to the pool")
+            else:
+                return getattr(self._item.transfer, attr)
+        return super().__getattribute__(attr)
+
+    @classmethod
+    def from_model(cls, model: StorageModel) -> BaseTransfer[StorageModel]:
+        raise InvalidTransferError("You should not call class methods on SafeTransfer instances")
+
+    def return_to_pool(self) -> None:
+        """Return the underlying transfer instance to the pool.
+
+        After this call all uses of this transfer instance will raise InvalidTransferError.
+
+        This method can be called multiple times.
+        """
+        if self._done:
+            return
+        self._pool_reclaim(self._item)
+        self._done = True
+        del self._item
+
+
+class TransferPool:
+    def __init__(self, *, max_pool_age: float = TRANSFER_CACHE_MAX_AGE, max_transfer_age: float = TRANSFER_MAX_AGE):
+        self._max_pool_age = max_pool_age
+        self._max_transfer_age = max_transfer_age
+        self._caches: dict[str, _TransferCache] = {}
+        self._is_thread_safe: dict[str, bool] = {}
+        self._mutex = threading.Lock()
+
+    @contextmanager
+    def with_transfer(self, storage_config: dict[str, Any]) -> Generator[BaseTransfer[StorageModel], None, None]:
+        """Yield a transfer object according to the provided storage_config.
+
+        The transfer object may be created or cached. While the context manager is running the transfer object
+        won't be available to other threads.
+
+        Users of this contextmanager should NOT use the transfer object outside the `with` block
+        """
+        transfer = self._get(storage_config)
+        yield transfer.transfer
+        self._put(transfer)
+
+    def get_transfer(self, storage_config: dict[str, Any]) -> SafeTransfer:
+        """Returns a transfer object.
+
+        The object returned is a proxy that will forward all accesses to the underlying transfer instance.
+        You MUST call `return_to_pool()` method once you are done with this transfer, and any access to the proxy
+        afterwards will raise an `InvalidTransferError`.
+
+        """
+        return SafeTransfer(item=self._get(storage_config), pool_reclaim=self._put)
+
+    def _prune_expired_caches(self) -> None:
+        self._caches = {key: pool for key, pool in self._caches.items() if not pool.is_expired()}
+
+    def _get(self, storage_config: dict[str, Any]) -> TransferCacheItem:
+        cache_key = _cache_key(storage_config)
+        with self._mutex:
+            if (cache := self._caches.get(cache_key)) and (cached_transfer := cache.get()):
+                return cached_transfer
+            transfer = rohmu_get_transfer(storage_config)
+            pool_cls = _TransferCacheForThreadSafeTransfer if transfer.is_thread_safe else _TransferCache
+            self._caches[cache_key] = pool_cls(max_age=self._max_pool_age)
```

Comment:
This could be really created only on put I guess (it would also eliminate the 'no transfer pool' branch there that is there now).